rfc/0019-hashed-mojmap.md

```diff
@@ -0,0 +1,97 @@
+# Hashed Mojmap
+
+## Summary
+
+This RFC proposes replacing intermediary with a hashed version of Mojmap. Everywhere where intermediary is currently used, hashed Mojmap will be used instead.
+
+## Motivation
+
+- One of the biggest, if not the biggest, reliance of Quilt on upstream Fabric is intermediary. For as long as Fabric keeps updating to new Minecraft versions, we will be behind them on updating because of this limitation. It is also not a good thing for a hard fork to have a reliance on upstream.
+
+- One of the primary goals of intermediary is to make sure that mods don't have to recompile between versions. However, when mods make the switch from Fabric to Quilt, they will have to recompile anyway. This gives us a unique opportunity to make breaking changes to the format of intermediary without breaking mods.
+
+- Intermediary was a good solution at the time of its inception. Since then, Mojang has provided us with their official mappings, with permission to use them as long as we don't redistribute them. Having to match intermediary between versions introduces a high maintenance cost for every snapshot and potentially also limits the bus factor of the whole project. The majority of the time updating Fabric to snapshots is typically spent matching.
+
+
+## Explanation
+
+Hashed Mojmap will produce base-62 hashes of information taken from Mojmap to produce mixed-case alphanumeric names like the following:
+
+- Classes: `Cls_hE89sr` (6 digits)
+- Fields: `fd_p9TXc40` (7 digits)
+- Methods: `md_fjH9ac8` (7 digits)
+
+To produce the hashes, the Mojmap names are processed as follows:
+
+1. Classes are converted from dotty format (`net.minecraft.a.Foo$Bar`) to slashy format (`net/minecraft/a/Foo$Bar`).
+1. Overriding methods in subclasses are discarded if present.
+1. Package names are removed from classes whose simple names are unique. That is, `net/minecraft/a/Bar` gets shortened to `Bar` if there is no other class named `Bar`, but if there are classes `net/minecraft/a/Foo` and `net/minecraft/b/Foo` that exist, then neither of them will be shortened. This includes references to these classes in method descriptors.
```

Comment:
I have to admit I am a little bit concerned about this.

If one day Mojang introduce a new class with the same name it'll break the old hash.

Sure it shouldn't happen often, but it's still something that's bugging me. Otherwise it's all good for me.

Reply:
This is a tradeoff.
The reasoning behind this was that package moves are more likely than adding classes with the same name.
The proposed system helps with package moves.

I'm also not convinced by this method, I think I'd be in favor of a more straightforward system that just hashes the entire name.

Reply:
> package moves are more likely than adding classes with the same name

They are quite a lot more likely according to my tests. This package stripping step helps *a lot* with breakages.